cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh

```diff
@@ -0,0 +1,297 @@
+          typename MultiLinestringRange2,
+          typename TempIt1,
+          typename TempIt2,
+          typename Offsets1,
+          typename Offsets2,
+          typename Offsets3,
+          typename IdRanges,
+          typename OutputIt1,
+          typename OutputIt2>
+void __global__ pairwise_linestring_intersection_simple(MultiLinestringRange1 multilinestrings1,
+                                                        MultiLinestringRange2 multilinestrings2,
+                                                        TempIt1 n_points_stored,
+                                                        TempIt2 n_segments_stored,
+                                                        Offsets1 num_points_offsets_first,
+                                                        Offsets2 num_segments_offsets_first,
+                                                        Offsets3 num_points_per_pair_first,
+                                                        IdRanges point_ids_range,
+                                                        IdRanges segment_ids_range,
+                                                        OutputIt1 points_first,
+                                                        OutputIt2 segments_first)
+{
+  using T       = typename MultiLinestringRange1::element_t;
+  using types_t = uint8_t;
+  using count_t = iterator_value_type<Offsets1>;
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < multilinestrings1.num_points();
+       idx += gridDim.x * blockDim.x) {
+    auto const part_idx = multilinestrings1.part_idx_from_point_idx(idx);
+    if (!multilinestrings1.is_valid_segment_id(idx, part_idx)) continue;
+    auto const lhs_linestring_idx = multilinestrings1.intra_part_idx(part_idx);
+    auto const lhs_segment_idx    = multilinestrings1.intra_point_idx(idx);
+    auto [a, b]                   = multilinestrings1.segment(idx);
+    auto const geometry_idx       = multilinestrings1.geometry_idx_from_part_idx(part_idx);
+    auto const multilinestring2   = multilinestrings2[geometry_idx];
+    auto const geometry_collection_offset =
+      num_points_offsets_first[geometry_idx] + num_segments_offsets_first[geometry_idx];
+
+    for (auto rhs_linestring_idx = 0; rhs_linestring_idx < multilinestring2.size();
+         ++rhs_linestring_idx) {
+      auto const linestring2 = multilinestring2[rhs_linestring_idx];
+      for (auto rhs_segment_idx = 0; rhs_segment_idx < linestring2.num_segments();
+           ++rhs_segment_idx) {
+        auto [c, d]                   = linestring2.segment(rhs_segment_idx);
+        auto [point_opt, segment_opt] = segment_intersection(segment<T>{a, b}, segment<T>{c, d});
+
+        // Writes geometry and origin IDs to output. Note that for each pair, intersecting
+        // points always precedes overlapping segments (arbitrarily).
+        if (point_opt.has_value()) {
+          auto r              = cuda::atomic_ref<count_t>{*(n_points_stored + geometry_idx)};
```

Comment:
Do you need the pointer arithmetic here?

```suggestion
          auto r              = cuda::atomic_ref<count_t>{n_points_stored[geometry_idx]};
```

Reply:
No specific reason, I think they are the same.

Reply:
I think clang-tidy would complain if and when we set it up. ;)

Reply:
I see - here I updated it with array accessor